scenario/clickactionbutton.go

```diff
@@ -196,109 +196,117 @@ func buttonActions(sessionState *session.State, actionState *action.State, obj *
+	actionType string, settings ActionSettings) error {
+
+	action := Action{
+		ActionCore: ActionCore{
+			Type:  ActionApplyBookmark,
+			Label: fmt.Sprintf("button-action-%s", actionType),
+		},
+		Settings: settings,
+	}
+	if isAborted, err := CheckActionError(action.Execute(sessionState, connectionSettings)); isAborted {
+		return errors.Wrapf(err, "%s was aborted", ActionClearAll)
+	} else if err != nil {
+		return errors.Wrapf(err, "%s failed", ActionClearAll)
+	}
+	return nil
+}
+
 // execute one action contained by a Sense action-button
-func (buttonAction *buttonAction) execute(sessionState *session.State, actionState *action.State) error {
+func (buttonAction *buttonAction) execute(sessionState *session.State, actionState *action.State,
+	connectionSettings *connection.ConnectionSettings, label string) error {
+
 	uplink := sessionState.Connection.Sense()
-	doc := DocWrapper{uplink.CurrentApp.Doc}
+	doc := uplink.CurrentApp.Doc
 	sendReq := func(f func(context.Context) error) error {
 		return sessionState.SendRequest(actionState, f)
 	}
+	executeContainerAction := func(actionType string, settings ActionSettings) error {
+		return executeContainerAction(sessionState, connectionSettings, actionType, settings)
+	}
 
 	switch buttonAction.ActionType {
-	case emptyAction: // do nothing
+	case emptyAction:
 		return nil
 
 	case unknownAction:
 		return errors.New("unknown button action")
 
+	// container button-actions:
+
 	case applyBookmark:
-		return sendReq(func(ctx context.Context) error {
-			success, err := doc.ApplyBookmark(ctx, buttonAction.Bookmark /*id*/)
-			if err != nil {
-				return errors.Wrapf(err, "error applying bookmark<%s>", buttonAction.Bookmark)
-			}
-			if !success {
-				return errors.Errorf("unsuccessful application bookmark<%s>", buttonAction.Bookmark)
-			}
-			return err
-		})
+		return executeContainerAction(ActionApplyBookmark,
+			&ApplyBookmarkSettings{
+				BookMarkSettings{
+					ID: buttonAction.Bookmark,
+				}})
+
+	case clearAllSelections:
+		return executeContainerAction(ActionClearAll, &ClearAllSettings{})
+
+	// TODO(atluq) the following buttonactions shall become container actions
+	// when implemetations of individual the actions exist.
+	// NOT container button-actions:
 
 	case moveBackwardsInSelections:
 		return sendReq(doc.Back)
 
 	case moveForwardsInSelections:
 		return sendReq(doc.Forward)
 
-	case clearAllSelections:
-		return sendReq(func(ctx context.Context) error {
-			return doc.ClearAll(ctx, buttonAction.SoftLock /*lockedAlso*/, "" /*stateName*/)
-		})
-
 	case clearSelectionsInOtherFields:
 		return sendReq(func(ctx context.Context) error {
-			field, err := uplink.FieldCache.LookupWithFallback(buttonAction.Field,
-				func(fieldName string) (*enigma.Field, error) {
-					return doc.GetField(ctx, fieldName)
-				},
-			)
+			field, err := fieldReq(doc.GetField).WithCache(&uplink.FieldCache).WithInputValidation()(ctx, buttonAction.Field, "")
```

Comment:
We haven't used much function chaining in the project before so something to be aware of. But the end result looks elegant in this use case. Nicely done!

Reply:
Thanks!

The enigma.Doc has a lot of functions that will be wrapped with the same logic in a lot of places. It Doc only were an interface, we could more easy wrap that one and reuse it in many places.